insights/parsers/grubenv.py

```diff
@@ -1,76 +1,60 @@
+-------------------------------------------------
 """
 from insights import get_active_lines, parser, Parser
 from insights.parsers import SkipException
 from insights.specs import Specs
 
 
-@parser(Specs.grubenv)
-class GrubEnv(Parser, dict):
+class GrubenvBase(Parser, dict):
     """
-    Parses the /boot/grub2/grubenv file and returns a dict
-    of the grubenv variables.
-
-    Sample output of the file::
+    This module parses the content of GRUB environment block file -
+    ``/boot/grub2/grubenv``. The file is laid out in a key=value format similar
+    to an ini file except it doesn't have any headers.
 
-        saved_entry=295e1ba1696e4fad9e062f096f92d147-4.18.0-305.el8.x86_64
-        kernelopts=root=/dev/mapper/root_vg-lv_root ro crashkernel=auto resume=/dev/mapper/root_vg-lv_swap rd.lvm.lv=root_vg/lv_root rd.lvm.lv=root_vg/lv_swap console=tty0 console=ttyS0,115200
-        boot_success=0
-        boot_indeterminate=2
-        tuned_params=transparent_hugepages=never
-        tuned_initrd=
+    The module stores the key/value pairs in itself which inherits a dict. This
+    file is only used in Grub 2, but in RHEL8 with BLS being the default. There
+    were several variables added that are referenced in the
+    ``/boot/loader/entries/*.conf`` files.
 
     Attributes:
         has_kernelopts (bool): Returns True/False depending on if kernelopts key is in the dict.
         kernelopts (bool): Returns the string of kernelopts from the dict.
         has_tuned_params (str): Returns True/False depending of if the tuned_params key is in the dict.
         tuned_params (str): Returns the string of tuned_params from the dict.
-
-    Examples:
-        >>> type(grubenv)
-        <class 'insights.parsers.grubenv.GrubEnv'>
-        >>> grubenv.has_kernelopts
-        True
-        >>> grubenv.kernelopts
-        'root=/dev/mapper/root_vg-lv_root ro crashkernel=auto resume=/dev/mapper/root_vg-lv_swap rd.lvm.lv=root_vg/lv_root rd.lvm.lv=root_vg/lv_swap console=tty0 console=ttyS0,115200'
-        >>> grubenv.has_tuned_params
-        True
-        >>> grubenv.tuned_params
-        'transparent_hugepages=never'
-        >>> grubenv['saved_entry']
-        '295e1ba1696e4fad9e062f096f92d147-4.18.0-305.el8.x86_64'
     """
+
     def __init__(self, context):
-        super(GrubEnv, self).__init__(context)
+        super(GrubenvBase, self).__init__(context)
 
     def parse_content(self, content):
         if not content:
             raise SkipException("Empty output.")
 
+        self._error = []
         data = dict()
         for line in get_active_lines(content):
             if "=" not in line:
+                self._error.append(line)
                 continue
 
             key, value = line.split("=", 1)
-
-            # Some keys can have empty values, so just skip them.
             if not value:
                 continue
 
             data[key] = value
 
-        if not data:
+        if (not data) and (not self._error):
```

Comment:
Could the check on errors be in a sub if statement, that way if there is no data but there were errors it should raise a ParseException, otherwise throw a SkipException. The way it is now, if there is no data parsed but an error line was stored it would return an empty dict when it shouldn't.

Reply:
Thanks for your comment. 
For 'GrubEnv' parser, it processes the content in '/boot/grub2/grubenv' file. The file saves environment variables for boot time, and will never contain error messages.
For 'Grub2EditenvList' parser, it parses the output of the 'grub2-editenv list' command.  The list the current variables in environment block file if there are no errors in the file configuration. Otherwise, there will be no data and only error messages in output. If there is no data parsed but an error line was stored, the return is not only an empty dict, but also self.error which is not empty. And error message is one the information we needed.
```
#e.g.
[root@rhel~]# grub2-editenv list
grub2-editenv: error: invalid environment block.
```

From my point of view, if it is just no data parsed, it is not necessary to throw an exception. If there is no data and no error, the parser must throw an exception.

Reply:
BTW, @TZ3070 - Because this is special processing, please add the above explanation to the code as a comment, to prevent someone from updating the parser in the future without understanding the logic.

Reply:
Oh, my fault. You have already added the comment.